File: protocol/clock.test.py
"""The clock contract, checked -- including that the check can fail.

A test that nothing can fail is not a test. That is not a general worry here,
it is the specific bug this project has already hit twice: a phrase origin
chosen by scoring boundaries against themselves, and a scores check that would
have skipped itself when its input went missing. So this asserts both halves --
that a measuring clock passes, and that a counting one does not.
"""
import sys
import os

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from clock import conformance, MeasuredClock, CountingClock, FakeDevice   # noqa: E402
from session import Session                                              # noqa: E402

out = []
def ok(name, cond, detail=""):
    out.append((bool(cond), name, detail))


# ---- a measuring clock keeps the contract ---------------------------------
for passed, name, detail in conformance(lambda d, n: MeasuredClock(d), "measured"):
    ok(name, passed, detail)

# ---- a counting clock does not, and the harness proves it -----------------
counting = conformance(lambda d, n: CountingClock(d, n), "counting")
failed = [r for r in counting if not r[0]]
ok("the harness can fail: a counting clock breaks the contract",
   len(failed) >= 3, f"{len(failed)} of {len(counting)} checks fail")
ok("and it fails on the three things that break a show",
   all(any(k in r[1] for r in failed)
       for k in ("ahead of a sound", "start delay changes", "slide apart")),
   "; ".join(r[1].split(": ", 1)[1] for r in failed))

# ---- the fixed-delay setting cannot rescue a counting clock ---------------
# Tune the delay perfectly at one start, then start somewhere else. This is the
# evening that keeps getting spent in the room.
t = [0.0]
now = lambda: t[0]
dev = FakeDevice(now, start_delay_s=0.080, rate=1.0)
c = CountingClock(dev, now)
c.play(0.0)
t[0] += 1.080
tuned = c.position() - dev.true_position()     # the delay that fixes THIS run

t2 = [0.0]
now2 = lambda: t2[0]
dev2 = FakeDevice(now2, start_delay_s=0.025, rate=1.0)
c2 = CountingClock(dev2, now2, latency_s=-tuned)   # same setting, next run
c2.play(0.0)
t2[0] += 1.025
left = (c2.position() - dev2.true_position()) * 1000
ok("a delay tuned on one run is wrong on the next", abs(left) > 20,
   f"tuned out {tuned*1000:.0f} ms, still {left:+.0f} ms out next time")

# ---- and the same setting on a measuring clock stays right ----------------
t3 = [0.0]
now3 = lambda: t3[0]
d3 = FakeDevice(now3, start_delay_s=0.025, rate=1.0)
m = MeasuredClock(d3)
m.play(0.0)
t3[0] += 1.025
ok("a measuring clock needs no tuning at all",
   abs(m.position() - d3.true_position()) * 1000 < 5,
   f"{(m.position() - d3.true_position())*1000:+.0f} ms with no setting")


# ---- the clock plugs into the protocol, and the music follows it ----------
score = {"grid": {"bpm": 120, "beats_per_bar": 4, "first_beat_s": 0.0, "first_bar": 1}}
t4 = [0.0]
now4 = lambda: t4[0]
d4 = FakeDevice(now4, start_delay_s=0.100, rate=1.0)
clock = MeasuredClock(d4)
s = Session(score, song_time=clock.position)
clock.play(0.0)
t4[0] += 0.050                       # half the start delay: no sound yet
ok("the protocol does not advance while the audio is still silent",
   s.now()["position"]["bar"] == 1 and s.now()["position"]["beat"] == 1,
   f"bar {s.now()['position']['bar']} beat {s.now()['position']['beat']}")

t4[0] += 2.100                       # 2.05 s of music at 120 bpm = bar 2 beat 1.1
pos = s.now()["position"]
ok("and lands on the bar the audio is actually at",
   pos["bar"] == 2 and abs(pos["beat"] - 1.1) < 0.01, f"bar {pos['bar']} beat {pos['beat']}")

# the same moment through a counting clock is a beat and a half early
t5 = [0.0]
now5 = lambda: t5[0]
d5 = FakeDevice(now5, start_delay_s=0.100, rate=1.0)
cc = CountingClock(d5, now5)
s5 = Session(score, song_time=cc.position)
cc.play(0.0)
t5[0] += 2.150
drift_beats = (s5.now()["seconds"] - s.now()["seconds"]) / (60.0 / 120)
ok("where a counting clock would have put it is measurably early",
   drift_beats > 0.15, f"{drift_beats:.2f} beats early")

bad = [r for r in out if not r[0]]
for passed, name, detail in out:
    if not passed:
        print(f"  FAIL  {name}   {detail}")
print(f"\n{len(bad)} of {len(out)} FAILED" if bad else f"\nall {len(out)} checks pass")
sys.exit(1 if bad else 0)

File: protocol/clock.test.js
/* The clock contract, checked -- including that the check can fail.
   ---------------------------------------------------------------------------
   A test nothing can fail is not a test. That is not a general worry here, it
   is the specific bug this project has hit three times: a phrase origin scored
   against itself, a scores check that would have skipped when its input went
   missing, and a bar base that agreed with whatever it was given. So this
   asserts both halves -- that the anchored clock keeps the contract, and that
   the other two break it, each in its own way. */
"use strict";
const { CountingClock, RawAudioClock, AnchoredClock, FakeAudio, conformance } = require("./clock.js");
const { Session } = require("./session.js");

const out = [];
const ok = (n, c, d) => out.push([!!c, n, d || ""]);

/* ---- the anchored clock keeps the contract ------------------------------ */
for (const [p, n, d] of conformance((a, o) => AnchoredClock(a, o), "anchored")) ok(n, p, d);

/* ---- and the other two fail, on different things ------------------------ */
const counting = conformance((a, o) => CountingClock(a, o), "counting");
const raw = conformance((a, o) => RawAudioClock(a, o), "raw");
const failed = l => l.filter(r => !r[0]).map(r => r[1].split(": ")[1]);

ok("the harness can fail: a counting clock is never accurate",
   failed(counting).length >= 3, failed(counting).join("; "));
ok("it fails on the three things that put a show out of sync",
   ["ahead of a sound", "start delay changes", "slide apart"]
     .every(k => failed(counting).some(f => f.includes(k))));
ok("a counting clock is smooth, which is why it is believed",
   !failed(counting).some(f => f.includes("judders")),
   "accurate never, smooth always -- it looks right and is not");

ok("reading the audio raw is accurate", 
   !["ahead of a sound", "slide apart", "start delay"].some(k => failed(raw).some(f => f.includes(k))));
ok("but it judders, which is the other way to get this wrong",
   failed(raw).some(f => f.includes("judders")), failed(raw).join("; "));

ok("only the anchored clock passes everything",
   conformance((a, o) => AnchoredClock(a, o)).every(r => r[0]));

/* ---- a calibration tuned on one run is wrong on the next ---------------- */
/* This is the evening that keeps getting spent in the room. */
{
  let t = 0; const now = () => t;
  const a = FakeAudio(now, { start_delay_s: 0.080, update_every_s: 0.001 });
  const c = CountingClock(a, { now });
  c.play(0); t += 1.08;
  const tuned = c.position() - a.trueTime();          /* fixes THIS run exactly */

  t = 0;
  const a2 = FakeAudio(now, { start_delay_s: 0.025, update_every_s: 0.001 });
  const c2 = CountingClock(a2, { now });
  c2.play(0); t += 1.025;
  const left = (c2.position() - tuned - a2.trueTime()) * 1000;
  ok("a calibration tuned on one run is wrong on the next", Math.abs(left) > 20,
     `tuned out ${(tuned * 1000).toFixed(0)} ms, still ${left.toFixed(0)} ms out next time`);
}

/* ---- the clock plugs into the protocol --------------------------------- */
const score = { grid: { bpm: 120, beats_per_bar: 4, first_beat_s: 0, first_bar: 1 } };
{
  let t = 0; const now = () => t;
  const a = FakeAudio(now, { start_delay_s: 0.100, update_every_s: 0.001 });
  const clock = AnchoredClock(a, { now });
  const s = Session(score, { songTime: () => clock.position() });
  clock.play(0);
  t += 0.050;                                    /* half the start delay: silent */
  const p0 = s.now().position;
  ok("the protocol does not advance while the audio is still silent",
     p0.bar === 1 && p0.beat === 1, `bar ${p0.bar} beat ${p0.beat}`);

  t += 2.100;                                    /* 2.05 s of music = bar 2 beat 1.1 */
  const p1 = s.now().position;
  ok("and lands on the bar the audio is actually at",
     p1.bar === 2 && Math.abs(p1.beat - 1.1) < 0.02, `bar ${p1.bar} beat ${p1.beat}`);
}

/* ---- what a rhythm game actually asks for ------------------------------ */
/* A note takes a fixed time to travel from the spawn point to the strike line.
   That travel time IS the lead time: ask the protocol what is coming in the
   next 1200 ms and those are exactly the notes to spawn this frame. */
{
  let t = 0; const now = () => t;
  const a = FakeAudio(now, { start_delay_s: 0.02, update_every_s: 0.050 });
  const clock = AnchoredClock(a, { now });
  const s = Session(score, { songTime: () => clock.position() });
  clock.play(0);
  t += 0.5;

  const APPROACH_MS = 1200;
  const spawned = new Map();                    /* one note per beat, never twice */
  const frame = 1 / 60;
  for (let i = 0; i < 600; i++) {               /* ten seconds of frames */
    t += frame;
    for (const b of s.next(APPROACH_MS)) {
      if (b.what) continue;                     /* section events, not beats */
      const key = `${b.bar}:${b.beat}`;
      if (!spawned.has(key)) spawned.set(key, { frame: i, in_ms: b.in_ms });
    }
  }
  const beats = [...spawned.values()];
  ok("every beat in the window gets spawned exactly once", spawned.size > 15,
     `${spawned.size} notes over ten seconds at 120 bpm`);

  /* The first call is different and a game has to expect it. next() answers
     "what is inside the window", not "what just entered it", so the first call
     after play or a seek hands back a backlog -- notes already part-way to the
     strike line, some of them nearly on it. Spawn those at their true distance
     or drop them; do not give them a full approach, or they arrive late. */
  const firstBatch = beats.filter(b => b.frame === 0).map(b => b.in_ms);
  ok("the first call hands back everything already in the window",
     firstBatch.length > 1 && Math.min(...firstBatch) < APPROACH_MS * 0.5,
     `${firstBatch.length} notes at ${Math.min(...firstBatch)}-${Math.max(...firstBatch)} ms`);

  const steady = beats.filter(b => b.frame > 0).map(b => b.in_ms);
  ok("after that, each new note arrives with a full approach of warning",
     steady.every(l => l > APPROACH_MS - 25 && l <= APPROACH_MS),
     `lead times ${Math.min(...steady)}-${Math.max(...steady)} ms, asked for ${APPROACH_MS}`);
}

/* ---- and a tempo change costs the game nothing ------------------------- */
{
  const s = Session(score, { now: () => 0 });
  s.seek(10.1).play();   /* not exactly on a beat: that is a degenerate case */
  const atOne = s.next(1200).filter(b => !b.what);
  s.rate(1.5);
  const atOneFive = s.next(1200).filter(b => !b.what);
  ok("a faster song puts more notes inside the same approach window",
     atOneFive.length > atOne.length, `${atOne.length} -> ${atOneFive.length} notes`);
  ok("and they are the same beats, arriving sooner",
     atOne[0].bar === atOneFive[0].bar && atOne[0].beat === atOneFive[0].beat
     && atOneFive[0].in_ms < atOne[0].in_ms,
     `bar ${atOne[0].bar} beat ${atOne[0].beat}: ${atOne[0].in_ms} -> ${atOneFive[0].in_ms} ms`);
}

for (const [p, n, d] of out) if (!p) console.log(`  FAIL  ${n}   ${d}`);
const bad = out.filter(r => !r[0]).length;
console.log(bad ? `\n${bad} of ${out.length} FAILED` : `\nall ${out.length} checks pass`);
process.exit(bad ? 1 : 0);
